provider/openstack/cinder.go

```diff
@@ -232,33 +238,53 @@ func (s *cinderVolumeSource) CreateVolumes(ctx context.ProviderCallContext, args
 	return results, nil
 }
 
-func (s *cinderVolumeSource) createVolume(arg storage.VolumeParams) (*storage.Volume, error) {
+func (s *cinderVolumeSource) createVolume(
+	ctx context.ProviderCallContext, arg storage.VolumeParams) (*storage.Volume, error) {
 	cinderConfig, err := newCinderConfig(arg.Attributes)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
+	// If this volume is being attached to an instance, attempt to provision
+	// the storage in the same availability zone.
+	// This helps to avoid a situation with all storage residing in a single
+	// AZ that upon failure would effectively take down attached instances
+	// whatever zone they were in.
+	az := ""
+	if arg.Attachment != nil {
+		instanceID := arg.Attachment.InstanceId
+		if instanceID != "" {
+			aZones, err := s.zonedEnv.InstanceAvailabilityZoneNames(ctx, []instance.Id{instanceID})
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if len(aZones) > 0 {
+				az = aZones[0]
+			}
```

Comment:
Suggest logging warning here.  Openstack compute has a default AZ of "nova", so aZones should never be empty.